File: problems/sum/services/sum_and_product_server.py
#!/usr/bin/env python3

from os import environ
import yaml
from sys import exit
from random import randrange

ENV_lang = environ["TAL_lang"]
ENV_numbers = environ["TAL_numbers"]
ENV_num_questions = int(environ["TAL_num_questions"])

with open("sum_and_product_server." + ENV_lang + ".yaml", 'r') as stream:
    try:
        messages_book = yaml.safe_load(stream)
    except yaml.YAMLError as exc:
        print(exc)

def print_lang(*args,**kwargs):
  message1, *other_messages = args
  msg1_code, *msg1_colors = message1
  msg1_str=eval(f"f'{messages_book[msg1_code]}'")
  msg1_code, *msg1_rendering = message1
  msg1_attrs = []
  if type(msg1_rendering[-1]) == list:
      msg1_attrs = msg1_rendering[-1]
      msg1_colors = msg1_rendering[:-1]
  else:
      msg1_colors = msg1_rendering      
  if len(other_messages)==0:
    if ENV_colored_feedback:
      cprint(msg1_str, *msg1_colors, attrs=msg1_attrs, **kwargs)
    else:
      print(msg1_str, **kwargs)
  else:
    if ENV_colored_feedback:
      cprint(msg1_str, *msg1_colors, attrs=msg1_attrs, end="")
    else:
      print(msg1_str, end="")  
    print_lang(*other_messages, **kwargs)



print_lang(["open-channel", "magenta", "on_blue"])        
#English: print(f"# I will serve: problem=sum, service=sum_and_product, numbers={ENV_numbers}, num_questions={ENV_num_questions}, colored_feedback={ENV_colored_feedback}, lang={ENV_lang}.")

gen_new_pair = True    
for _ in range(ENV_num_questions):
    if gen_new_pair:
        if ENV_numbers == "onedigit":
            x = randrange(10)
            y = randrange(10)
        elif ENV_numbers == "twodigits":
            x = randrange(100)
            y = randrange(100)
        else:
            x = randrange(2**32)
            y = randrange(2**32)
    cprint("? {x+y} {x*y}", "yellow", "on_blue", ["bold"])
    spoon = input().strip()
    while spoon[0] == '#':
        cprint(spoon, 'magenta', 'on_blue')
        spoon = input().strip()
    a, b = map(int, spoon.split(" "))
    gen_new_pair = False
    if a+b > x+y:
       print_lang(["no_answ", "red", "on_blue", ["blink"]], ["over-sum", "yellow", "on_blue", ["underline"]])        
       #English: print(f"No! indeed, {a}+{b}={a+b} > {x+y}.")
    elif a+b < x+y:    
       print_lang(["no_answ", "red", "on_blue", ["blink"]], ["under-sum", "yellow", "on_blue", ["underline"]])        
       #English: print(f"No! indeed, {a}+{b}={a+b} < {x+y}.")
    elif a*b > x*y:    
       print_lang(["no_answ", "red", "on_blue", ["blink"]], ["over-product", "yellow", "on_blue", ["underline"]])        
       #English: print(f"No! indeed, {a}*{b}={a*b} > {x*y}.")
    elif a*b < x*y:    
       print_lang(["no_answ", "red", "on_blue", ["blink"]], ["under-product", "yellow", "on_blue", ["underline"]])        
       #English: print(f"No! indeed, {a}*{b}={a*b} < {x*y}.")
    else:
        assert (a + b == x+y) and (a * b == x*y)
        print_lang(["ok_answ", "green", "on_blue"], ["ok", "grey", "on_blue"])        
        #English: print(f"OK! indeed, {a}+{b}={x+y} and {a}*{b}={x*y}.")
        gen_new_pair = True

print_lang(["got-bored", "magenta", "on_blue"])        
#English: print("! (I got bored)")
exit(0)

File: example_problems/pure_java/sum/services/sum_and_product/simple/sum_and_product_server.java
import java.util.Scanner;
import java.util.Map;
import java.util.Random;

public class sum_and_product_server {

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        String spoon = "";
        String numbers = "";
        int num_question=0;
        int x = 0;
        int y = 0;
        Random randrange = new Random();

        Map <String, String> map = System.getenv();
        for (Map.Entry <String, String> entry: map.entrySet()) {

            if(entry.getKey().equals("TAL_NUMBERS"))
                numbers= entry.getValue();
            if(entry.getKey().equals("TAL_NUM_QUESTIONS"))
                num_question= Integer.parseInt(entry.getValue());

        }
        System.out.println(numbers);
        System.out.println(num_question);
        boolean gen_new_pair = true;
        for (int i=0;i<num_question;i++) {
            if(numbers.equals("onedigit")) {
                x= randrange.nextInt(10);
                y=randrange.nextInt(10);
            }
            else if(numbers.equals("twodigits")) {
                x=randrange.nextInt(100);
                y=randrange.nextInt(100);
            }
            else {
                x=randrange.nextInt(2*32);
                y=randrange.nextInt(2*32);
            }
            int sum = x+y;
            int prod=x*y;
            System.out.println("? "+sum + " "+prod);
            spoon = scan.nextLine();
            while(spoon.charAt(0)=='#') {
                spoon=scan.nextLine();
            }
            int a = Integer.parseInt(spoon.split(" ")[0]);
            int b = Integer.parseInt(spoon.split(" ")[1]);
            gen_new_pair = false;
            if (a+b > x+y)
                System.out.println("No! indeed,"+a+ "+"+b+"="+(a+b)+">"+(x+y)+".");
            else if (a+b < x+y)
                System.out.println("No! indeed,"+a+ "+"+b+"="+(a+b)+"<"+(x+y)+".");
            else if (a*b > x*y)
                System.out.println("No! indeed,"+a+ "*"+b+"="+(a*b)+">"+(x*y)+".");
            else if (a*b < x*y)
                System.out.println("No! indeed,"+a+ "*"+b+"="+(a*b)+">"+(x*y)+".");
            else {
                System.out.println("Ok! indeed,"+a+ "+"+b+"="+(a+b)+" and"+a+"*"+b+"="+(a*b)+".");
                gen_new_pair = false;
            }

        }
        System.exit(0);
    }
}
    /*
    if gen_new_pair:
        if ENV['numbers'] == "onedigit":
            x = randrange(10)
            y = randrange(10)
        elif ENV['numbers'] == "twodigits":
            x = randrange(100)
            y = randrange(100)
        else:
            x = randrange(2**32)
            y = randrange(2**32)
    print(f"? {x+y} {x*y}")
        System.exit(0);

    }
}

/*


        while (true) {
            spoon = scan.nextLine();
            while (!(spoon.charAt(0) == '?')) {
                spoon = scan.nextLine();
            }
            int s = Integer.parseInt(spoon.split(" ")[1]);
            int p = Integer.parseInt(spoon.split(" ")[2]);
	    int delta = (int) Math.sqrt(s * s - 4 * p);
	    int x1 = (s - delta) / 2;
	    int x2 = s - x1;
	    System.out.println(x1 + " " + x2);
        }
    }
}

ENV={}
ENV['numbers'] = environ["TAL_numbers"]
ENV['num_questions'] = int(environ["TAL_num_questions"])

print(f"# I will serve: problem=sum, service=sum_and_product, numbers={ENV['numbers']}, num_questions={ENV['num_questions']}.")

gen_new_pair = True
for _ in range(ENV['num_questions']):
    if gen_new_pair:
        if ENV['numbers'] == "onedigit":
            x = randrange(10)
            y = randrange(10)
        elif ENV['numbers'] == "twodigits":
            x = randrange(100)
            y = randrange(100)
        else:
            x = randrange(2**32)
            y = randrange(2**32)
    print(f"? {x+y} {x*y}")
    spoon = input().strip()
    while spoon[0] == '#':
        spoon = input().strip()
    a, b = map(int, spoon.split(" "))
    gen_new_pair = False
    if a+b > x+y:
        print(f"No! indeed, {a}+{b}={a+b} > {x+y}.")
    elif a+b < x+y:
        print(f"No! indeed, {a}+{b}={a+b} < {x+y}.")
    elif a*b > x*y:
        print(f"No! indeed, {a}*{b}={a*b} > {x*y}.")
    elif a*b < x*y:
        print(f"No! indeed, {a}*{b}={a*b} < {x*y}.")
    else:
        assert (a + b == x+y) and (a * b == x*y)
        print(f"Ok! indeed, {a}+{b}={x+y} and {a}*{b}={x*y}.")
        gen_new_pair = True

exit(0)
*/

/*


        while (true) {
            spoon = scan.nextLine();
            while (!(spoon.charAt(0) == '?')) {
                spoon = scan.nextLine();
            }
            int s = Integer.parseInt(spoon.split(" ")[1]);
            int p = Integer.parseInt(spoon.split(" ")[2]);
	    int delta = (int) Math.sqrt(s * s - 4 * p);
	    int x1 = (s - delta) / 2;
	    int x2 = s - x1;
	    System.out.println(x1 + " " + x2);
        }
    }
}

ENV={}
ENV['numbers'] = environ["TAL_numbers"]
ENV['num_questions'] = int(environ["TAL_num_questions"])

print(f"# I will serve: problem=sum, service=sum_and_product, numbers={ENV['numbers']}, num_questions={ENV['num_questions']}.")

gen_new_pair = True    
for _ in range(ENV['num_questions']):
    if gen_new_pair:
        if ENV['numbers'] == "onedigit":
            x = randrange(10)
            y = randrange(10)
        elif ENV['numbers'] == "twodigits":
            x = randrange(100)
            y = randrange(100)
        else:
            x = randrange(2**32)
            y = randrange(2**32)
    print(f"? {x+y} {x*y}")
    spoon = input().strip()
    while spoon[0] == '#':
        spoon = input().strip()
    a, b = map(int, spoon.split(" "))
    gen_new_pair = False
    if a+b > x+y:
        print(f"No! indeed, {a}+{b}={a+b} > {x+y}.")
    elif a+b < x+y:    
        print(f"No! indeed, {a}+{b}={a+b} < {x+y}.")
    elif a*b > x*y:    
        print(f"No! indeed, {a}*{b}={a*b} > {x*y}.")
    elif a*b < x*y:    
        print(f"No! indeed, {a}*{b}={a*b} < {x*y}.")
    else:
        assert (a + b == x+y) and (a * b == x*y)
        print(f"Ok! indeed, {a}+{b}={x+y} and {a}*{b}={x*y}.")
        gen_new_pair = True

exit(0)
*/
